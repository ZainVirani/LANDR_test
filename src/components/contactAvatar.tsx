import { Avatar, Typography } from '@mui/material';
import React from 'react';
import { Contact } from '../types/contact';
import { getAvatarColor } from '../utils/avatarColors';

const avatarSizeStyle = { width: '75px', height: '75px' };

export const ContactAvatar = ({ contact }: { contact: Contact }) => {
  const contactAvatar = contact.profilePhotoLocation ? (
    <Avatar
      className="contactPhoto"
      alt={contact.name}
      src={contact.profilePhotoLocation}
      sx={avatarSizeStyle}
    />
  ) : (
    <Avatar
      className="contactPhoto"
      alt={contact.name}
      sx={{ ...avatarSizeStyle, ...getAvatarColor(contact.avatarColorKey) }}>
      <Typography variant="h4">{contact.name.slice(0, 1)}</Typography>
    </Avatar>
  );
  return (
    <div className="contactAvatar">
      {contactAvatar}
      <Typography noWrap className="contactName" variant="h6" component="h6">
        {contact.name}
      </Typography>
    </div>
  );
};
