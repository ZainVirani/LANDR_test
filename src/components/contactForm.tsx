import React from 'react';
import { Formik, Form, Field, ErrorMessage } from 'formik';
import { ContactFromForm } from '../types/contact';
import { Button, Typography } from '@mui/material';

export const ContactForm = ({
  initialValues,
  onSubmit,
  submitButtonText
}: {
  initialValues: Partial<ContactFromForm>;
  onSubmit: (values: Partial<ContactFromForm>) => Promise<void>;
  submitButtonText: string;
}) => {
  return (
    <Formik
      initialValues={{
        name: '',
        jobTitle: '',
        address: '',
        phoneNumbers: '',
        email: '',
        ...initialValues
      }}
      validate={(values) => {
        const errors: Record<string, string> = {};
        if (!values.name) {
          errors.name = 'Required';
        }
        if (
          values.email &&
          !/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i.test(values.email)
        ) {
          errors.email = 'Invalid email address';
        }
        return errors;
      }}
      onSubmit={onSubmit}>
      {({ isSubmitting }) => (
        <Form className="contactForm">
          <label className="formLabel"><Typography>{'Name (required)'}</Typography></label>
          <Field focus name="name" placeholder="Name" />
          <ErrorMessage name="name" component="div" />
          <label className="formLabel"><Typography>Job Title</Typography></label>
          <Field name="jobTitle" placeholder="Job Title" />
          <ErrorMessage name="jobTitle" component="div" />
          <label className="formLabel"><Typography>Address</Typography></label>
          <Field name="address" placeholder="Address" />
          <ErrorMessage name="address" component="div" />
          <label className="formLabel"><Typography>{'Phone Numbers (comma separated)'}</Typography></label>
          <Field
            name="phoneNumbers"
            placeholder="Phone Numbers (comma separated)"
          />
          <ErrorMessage name="phoneNumbers" component="div" />
          <label className="formLabel"><Typography>Email</Typography></label>
          <Field type="email" name="email" placeholder="Email" />
          <ErrorMessage name="email" component="div" />
          <Button type="submit" disabled={isSubmitting} variant="outlined">
            <Typography>{submitButtonText}</Typography>
          </Button>
        </Form>
      )}
    </Formik>
  );
};
