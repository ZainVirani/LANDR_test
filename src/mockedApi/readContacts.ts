export async function readContacts() {
  const data = await fetch('http://localhost:3000/contacts');
  return data.json();
}
