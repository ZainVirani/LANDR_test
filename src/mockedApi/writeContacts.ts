import { v4 as uuidv4 } from 'uuid';
import { Contact, NewContact } from '../types/contact';

export async function addContact(newContact: NewContact) {
  await fetch('http://localhost:3000/contacts', {
    method: 'POST',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      ...newContact,
      id: uuidv4()
    })
  });
}

export async function editContact(contactId: string, contact: Contact) {
  await fetch(`http://localhost:3000/contacts/${contactId}`, {
    method: 'PUT',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(contact)
  });
}

export async function deleteContact(contactId: string) {
  await fetch(`http://localhost:3000/contacts/${contactId}`, {
    method: 'DELETE',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json'
    }
  });
}
