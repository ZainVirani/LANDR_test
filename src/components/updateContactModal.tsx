import { Button, Modal, Typography } from '@mui/material';
import { Box } from '@mui/system';
import React from 'react';
import {
  addContact,
  deleteContact,
  editContact
} from '../mockedApi/writeContacts';
import { Contact, ContactFromForm } from '../types/contact';
import { getRandomAvatarColor } from '../utils/avatarColors';
import { ContactForm } from './contactForm';

export const UpdateContactModal = ({
  initialValues,
  open,
  refetchData,
  handleModalClose
}: {
  initialValues?: Contact;
  open: boolean;
  refetchData: () => void;
  handleModalClose: () => void;
}) => {
  if (!initialValues) {
    return (
      <Modal open={open} onClose={handleModalClose}>
        <Box className="contactModal">
          <Typography>Error: selected user is empty</Typography>
        </Box>
      </Modal>
    );
  }

  const validateAndAddContact = async (values: ContactFromForm) => {
    await editContact(initialValues.id, {
      ...values,
      id: initialValues.id,
      phoneNumbers: values.phoneNumbers.replace(' ', '').split(','),
      profilePhotoLocation: undefined
    });
    handleModalClose();
    refetchData();
  };

  const deleteContactWrapper = async () => {
    await deleteContact(initialValues.id);
    handleModalClose();
    refetchData();
  };

  const formattedInitialValues: ContactFromForm = {
    ...initialValues,
    phoneNumbers: initialValues.phoneNumbers
      ? initialValues.phoneNumbers.join()
      : ''
  };

  return (
    <Modal open={open} onClose={handleModalClose}>
      <Box className="contactModal">
        <ContactForm
          initialValues={formattedInitialValues}
          onSubmit={validateAndAddContact}
          submitButtonText={'Edit Contact'}
        />
        <Button className="deleteUser" onClick={deleteContactWrapper} variant="outlined" color="error">
          <Typography>Delete User</Typography>
        </Button>
      </Box>
    </Modal>
  );
};
