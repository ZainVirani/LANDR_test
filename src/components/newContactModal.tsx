import { Modal } from '@mui/material';
import { Box } from '@mui/system';
import React from 'react';
import { addContact } from '../mockedApi/writeContacts';
import { ContactFromForm } from '../types/contact';
import { getRandomAvatarColor } from '../utils/avatarColors';
import { ContactForm } from './contactForm';

export const NewContactModal = ({
  open,
  refetchData,
  handleModalClose
}: {
  open: boolean;
  refetchData: () => void;
  handleModalClose: () => void;
}) => {
  const validateAndAddContact = async (values: ContactFromForm) => {
    await addContact({
      ...values,
      phoneNumbers: values.phoneNumbers.replace(' ', '').split(','),
      profilePhotoLocation: undefined,
      avatarColorKey: getRandomAvatarColor()
    });
    handleModalClose();
    refetchData();
  };

  return (
    <Modal open={open} onClose={handleModalClose}>
      <Box className="contactModal">
        <ContactForm
          initialValues={{}}
          onSubmit={validateAndAddContact}
          submitButtonText={'Add Contact'}
        />
      </Box>
    </Modal>
  );
};
