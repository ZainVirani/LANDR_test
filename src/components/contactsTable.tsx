import React, { useState, useEffect } from 'react';
import Button from '@mui/material/Button';
import { Contact } from '../types/contact';
import { readContacts } from '../mockedApi/readContacts';
import { ContactAvatar } from './contactAvatar';
import { Box, Grid, Typography } from '@mui/material';
import { NewContactModal } from './newContactModal';
import { UpdateContactModal } from './updateContactModal';

export const ContactsTable = () => {
  const [contacts, setContacts] = useState<Contact[]>([]);
  const [newContactModalOpen, setNewContactModalOpen] = useState(false);
  const [updateContactModalOpen, setUpdateContactModalOpen] = useState(false);
  const [selectedContact, setSelectedContact] = useState<Contact | undefined>(
    undefined
  );

  const handleNewContactModalOpen = () => setNewContactModalOpen(true);
  const handleNewContactModalClose = () => setNewContactModalOpen(false);
  const handleUpdateContactModalOpen = (contact: Contact) => {
    setSelectedContact(contact);
    setUpdateContactModalOpen(true);
  };
  const handleUpdateContactModalClose = () => setUpdateContactModalOpen(false);

  useEffect(() => {
    async function readContactsData() {
      let response = await readContacts();
      setContacts(response);
    }
    readContactsData();
  }, []);

  const refetchData = async () => {
    let newContactsList = await readContacts();
    setContacts(newContactsList);
  };

  return (
    <Box className="contactsTable">
      <NewContactModal
        open={newContactModalOpen}
        refetchData={refetchData}
        handleModalClose={handleNewContactModalClose}
      />
      <UpdateContactModal
        initialValues={selectedContact}
        open={updateContactModalOpen}
        refetchData={refetchData}
        handleModalClose={handleUpdateContactModalClose}
      />
      <Typography className="mainTitle" variant="h4" component="h4">
        LANDR Contacts
      </Typography>
      <Box className="addContactBox">
        <Button onClick={handleNewContactModalOpen}>Add Contact</Button>
      </Box>
      <Grid className="contactsGrid" container spacing={2}>
        {contacts.map((contact) => (
          <Grid
            key={contact.id}
            onClick={() => handleUpdateContactModalOpen(contact)}
            className="contactGridItem"
            item
            xs={4}>
            <Box className="contactGridBox">
              <ContactAvatar contact={contact} />
            </Box>
          </Grid>
        ))}
      </Grid>
    </Box>
  );
};
