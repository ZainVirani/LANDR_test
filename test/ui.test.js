import 'jsdom-global/register';
import React from 'react';
import Enzyme, { shallow, mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json'
import { ContactAvatar } from '../src/components/contactAvatar';
import { ContactsTable } from '../src/components/ContactsTable';
import { NewContactModal } from '../src/components/newContactModal';
import { UpdateContactModal } from '../src/components/updateContactModal';
import * as apiRead from '../src/mockedApi/readContacts';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

Enzyme.configure({ adapter: new Adapter() });

const fakeUser = {
  id: "0",
  name: 'Zain Virani',
  jobTitle: 'TBD',
  address: 'Toronto',
  phoneNumbers: ['514-715-6639'],
  email: 'zainvirani96@gmail.com',
  avatarColorKey: 0,
};

const fakeNewUser = {
  id: "2",
  name: 'Vain Zirani',
  jobTitle: '',
  address: 'Montreal',
  phoneNumbers: ['514-715-9936'],
  email: 'notme@gmail.com',
  avatarColorKey: 1,
}

const userList = [fakeUser]

apiRead.readContacts = jest.fn(() => userList)

describe('Contact Avatar', () => {
  it('should render a user\'s avatar', () => {
    const avatar = shallow(<ContactAvatar
      contact={fakeUser}
    />);
    expect(shallowToJson(avatar)).toMatchSnapshot();
  })
});

// describe('Contacts Table', () => {
//   it('should render 1 user', () => {
//     const table = mount(<ContactsTable />);
//     expect(shallowToJson(table)).toMatchSnapshot();
//   })
//   it('should render 2 users', () => {
//     userList.push(fakeNewUser)
//     const table = mount(<ContactsTable />);
//     expect(shallowToJson(table)).toMatchSnapshot();
//   })
// });

// describe('New Contact Modal', () => {
//   let root;
//   afterEach(() => {
//     root.unmount();
//   })
//   it('should attempt to write data', () => {
//     act(() => {
//       root = create(<NewContactModal />)
//     })
//     const instance = root.root;
//     expect(root.toJSON()).toMatchSnapshot();
//     console.log(root.toTree())
//     const buttons = instance.findByType('button');
//     act(() => {
//     })
//     console.log(buttons);
//   })
// });

// describe('Update Contact Modal', () => {
//   let root;
//   afterEach(() => {
//     root.unmount();
//   })
//   it('should attempt to update data', () => {
//     act(() => {
//       root = create(<UpdateContactModal />)
//     })
//     const instance = root.root;
//     expect(root.toJSON()).toMatchSnapshot();
//     const buttons = instance.findByType('button');
//     act(() => {
//     })
//     console.log(buttons);
//   })
//   it('should attempt to delete data', () => {
//     act(() => {
//       root = create(<UpdateContactModal />)
//     })
//     const instance = root.root;
//     expect(root.toJSON()).toMatchSnapshot();
//     const buttons = instance.findByType('button');
//     act(() => {
//     })
//     console.log(buttons);
//   })
// });