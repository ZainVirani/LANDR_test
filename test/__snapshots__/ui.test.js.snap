// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact Avatar should render a user's avatar 1`] = `
<div
  className="contactAvatar"
>
  <ForwardRef(Avatar)
    alt="Zain Virani"
    className="contactPhoto"
    sx={
      {
        "bgcolor": "#ff5722",
        "height": "75px",
        "width": "75px",
      }
    }
  >
    <ForwardRef(Typography)
      variant="h4"
    >
      Z
    </ForwardRef(Typography)>
  </ForwardRef(Avatar)>
  <ForwardRef(Typography)
    className="contactName"
    component="h6"
    noWrap={true}
    variant="h6"
  >
    Zain Virani
  </ForwardRef(Typography)>
</div>
`;
